server/src-lib/Hasura/RQL/DDL/Metadata.hs

```diff
@@ -142,11 +146,16 @@ applyQP2 (ReplaceMetadata _ tables functionsMeta
 
   withPathK "tables" $ do
     -- tables and views
-    indexedForM_ tables $ \tableMeta -> do
-      let tableName = tableMeta ^. tmTable
-          isEnum = tableMeta ^. tmIsEnum
-          config = tableMeta ^. tmConfiguration
-      void $ Schema.trackExistingTableOrViewP2 tableName isEnum config
+    
+    -- NOTE: this is essentially a more efficient bulk 'Schema.trackExistingTableOrViewP2':
+    withBuildSchemaCache $ do
+      typeMap <- GC._gTypes . scDefaultRemoteGCtx <$> askSchemaCache
+      indexedForM tables $ \TableMeta{..} -> do
+        -- It should be okay to do this check against old cache as we work through new tables; we've 
+        -- already checked for duplicates in the track table payload:
+        GS.checkConflictingNode typeMap $ GS.qualObjectToName _tmTable
+        saveTableToCatalog _tmTable _tmIsEnum _tmConfiguration
+        return (MOTable _tmTable)
```

Comment:
I think we don't need to build schema cache here and check for inconsistent tables/views. Ideally in replace metadata, we should be _only saving the metadata in the catalog_ and do `buildSchemaCacheStrict` at the end. 

Define a new function with name like `saveMetadata` with definition as follows
```haskell
saveMetadata :: (MonadTx m, HasSystemDefined m) => ReplaceMetadata -> m ()
```
It should only insert the metadata items in the catalog. Move the code from `withPathK "tables"` to `forM_ actions .. ` snippet to this function. Also, we should be only using `addRemoteSchemaToCatalog` instead of `addRemoteSchemaP2` for remote schemas. Eventually, `applyQP2` function will look like

```haskell
applyQP2 replaceMetadata = do
  liftTx clearMetadata
  saveMetadata replaceMetadata
  buildSchemaCacheStrict
  pure successMsg
```

Reply:
Thanks, Rakesh. Do you think it makes more sense for me to reassign this issue to you? Or maybe to open a new issue that is about a bigger refactoring of replace_metadata?

I'm happy to attempt what you're describing, but with the state of documentation and my unfamiliarity with this part of the codebase I wonder if that makes sense. 

Let me know what you think.

Reply:
I agree with @rakeshkky—I don’t think we should worry about building the schema cache at all until after we’ve dropped/recreated all the catalog information. I think what you have is mostly fine, you can just drop `withBuildSchemaCache`. We rebuild the schema cache again once we’re done, anyway.

Reply:
@jberryman No worries, I'll push a commit for the changes I requested. Please make sure I have write permissions to your fork. :)

Reply:
Cool! I invited you as a "collaborator" @rakeshkky . Hopefully that did it.

Appreciate the attention here; with all that code being in `CacheRWM` it took a lot of staring to decide that the small change I made was probably okay. It seemed as though there might be many places where we do some validation against what would be a stale schema cache, but like I said I'm still getting acquainted with stuff here.

Reply:
Yeah, there’s a lot going on. I’m sure `replace_metadata` could be cleaned up, but I don’t think it absolutely has to be done right now.

One thing that stands out to me is that we call `buildSchemaCacheStrict` after dropping the old metadata but before creating the new metadata. This has the effect of purging the metadata cache, so even if the new metadata is mostly the same as the old one, it will reprocess everything.

I’d hesitate to just delete the call without ensuring some of this code isn’t incorrectly relying on it, but it’s something we could improve if we decide to clean things up.

Reply:
@jberryman I've pushed a couple of commits. Please test this against the table set of ~1000 and confirm the replace metadata is taking minimal time.

Reply:
@lexi-lambda
If we do `buildSchemaCacheStrict` only at the end of dropping metadata and creating the new metadata, the process of building schema cache will be totally **incremental**! And that is how the "v2 migrations workflow" meant to be which uses the `replace_metadata` API. Now the API will be even more faster. 

Using CLI v2-migrations flow:
I have 1000 tables before and I'll edit my metadata file to track one more table. I'll apply the metadata, then server only builds the cache for the extra table since it is new and older tables' metadata ain't changed. 
I would like to know your thoughts. :slightly_smiling_face:

Reply:
I very much agree! The current approach of calling `buildSchemaCacheStrict` after dropping the old metadata is quite a bad deoptimization, and doing things incrementally is much better.

Sadly, we’ll still pay a nontrivial cost on every `replace_metadata` call, since we still have to *load* the whole schema into memory from Postgres. That query will take time, reading and parsing it into Haskell will take time, and checking what has/has not changed will take time. Still, I expect it to be an improvement! I just don’t know how high the remaining overhead will be.